spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/AvoidPreviousInstanceRoundRobinLoadBalancer.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2012-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.loadbalancer.core;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import reactor.core.publisher.Mono;
+
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.cloud.client.ServiceInstance;
+import org.springframework.cloud.client.loadbalancer.DefaultResponse;
+import org.springframework.cloud.client.loadbalancer.EmptyResponse;
+import org.springframework.cloud.client.loadbalancer.Request;
+import org.springframework.cloud.client.loadbalancer.Response;
+import org.springframework.cloud.client.loadbalancer.RetryLoadBalancerInterceptor;
+import org.springframework.cloud.client.loadbalancer.RetryableRequestContext;
+
+/**
+ * A modified {@link RoundRobinLoadBalancer} implementation that avoids picking the same
+ * service instance while retrying requests.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 3.0.0
+ * @see RetryLoadBalancerInterceptor
+ */
+public class AvoidPreviousInstanceRoundRobinLoadBalancer extends RoundRobinLoadBalancer {
```

Comment:
I wonder if we can do this without tying ourselves to an algorithm?

Reply:
It comes down to getting a `List<ServiceInstance>` eventually. What you've implemented is basically a `ServiceInstanceListSupplier.get(Request)` or maybe `.get(RequestContext)`.

Reply:
I think that would work.

in `ServiceInstanceListSupplier`

```
  // for backwards compatibility
	default Flux<List<ServiceInstance>> get(Request<?> request) {
		return get();
	}
```

in a new `RetryAwareServiceInstanceListSupplier`
```
	public Flux<List<ServiceInstance>> get(Request req) {
		get().map(instanceList -> {
			//... this is the logic below from choose() and getInstanceResponse()
		});
	}
```
Then we can add `RetryAwareServiceInstanceListSupplier` to the builder and for the defaults where retry is used we automatically wrap the existing `ServiceInstanceListSupplier`

Reply:
Ok. Will work on changing it that way.